bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java

```diff
@@ -0,0 +1,805 @@
+import org.openhab.io.transport.modbus.ModbusReadFunctionCode;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegister;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.openhab.io.transport.modbus.ModbusWriteRegisterRequestBlueprint;
+import org.openhab.io.transport.modbus.endpoint.ModbusSlaveEndpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+/**
+ * The {@link HeliosEasyControlsHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Bernhard Bauer - Initial contribution
+ */
+@NonNullByDefault
+public class HeliosEasyControlsHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(HeliosEasyControlsHandler.class);
+
+    private @Nullable HeliosEasyControlsConfiguration config;
+
+    private @Nullable ScheduledFuture<?> pollingJob;
+
+    private @Nullable Map<String, HeliosVariable> variableMap;
+
+    /**
+     * This flag is used to ensure read requests (consisting of a write and subsequent read) are not influenced by
+     * another transaction
+     */
+    private Map<ModbusSlaveEndpoint, Semaphore> transactionLocks = new ConcurrentHashMap<>();
+
+    private Gson gson = new Gson();
+
+    private @Nullable ModbusCommunicationInterface comms;
+
+    private int dateFormat = -1;
+    private ZonedDateTime sysDate = ZonedDateTime.now(); // initialize with local system time as a best guess
+                                                         // before reading from device
+
+    private class BypassDate {
+        // initialization to avoid issues when updating before all variables were read
+        private int month = 1;
+        private int day = 1;
+
+        public void setMonth(int month) {
+            this.month = month;
+        }
+
+        public void setDay(int day) {
+            this.day = day;
+        }
+
+        public DateTimeType toDateTimeType() {
+            return new DateTimeType(ZonedDateTime.of(1900, this.month, this.day, 0, 0, 0, 0, ZoneId.of("UTC+00:00")));
+        }
+    }
+
+    private @Nullable BypassDate bypassFrom, bypassTo;
+
+    public HeliosEasyControlsHandler(Thing thing) {
+        super(thing);
+    }
+
+    /**
+     * Reads variable definitions from JSON file and store them in variableMap
+     */
+    private void readVariableDefinition() {
+        Type vMapType = new TypeToken<Map<String, HeliosVariable>>() {
+        }.getType();
+        try {
+            InputStreamReader jsonFile = new InputStreamReader(
+                    getClass().getResourceAsStream(HeliosEasyControlsBindingConstants.VARIABLES_DEFINITION_FILE));
+            BufferedReader reader = new BufferedReader(jsonFile);
+            this.variableMap = gson.fromJson(reader, vMapType);
+            reader.close();
+            jsonFile.close();

```

Comment:
Closing the streams must be done in the `finally` clause to ensure they are closed also if an exception is raised. Or you could use Java's auto-closeable feature:
```suggestion
        try (InputStreamReader jsonFile = new InputStreamReader(
                getClass().getResourceAsStream(HeliosEasyControlsBindingConstants.VARIABLES_DEFINITION_FILE));
                BufferedReader reader = new BufferedReader(jsonFile)) {
            this.variableMap = gson.fromJson(reader, vMapType);
```